src/replication.c

```diff
@@ -788,48 +869,136 @@ update_replica_entry(spec_t *spec, replica_t *replica, int iofd, char *replicaip
 	write(replica->iofd, rio, sizeof(zvol_io_hdr_t));
 	write(replica->iofd, spec->lu->volname, strlen(spec->lu->volname));
 	free(rio);
-	rio = NULL;
+
 	update_volstate(spec);
-	struct epoll_event event;
+
 	event.data.fd = replica->iofd;
 	event.events = EPOLLIN | EPOLLET;
 	rc = epoll_ctl(spec->receiver_epfd, EPOLL_CTL_ADD, iofd, &event);
 	if (rc == -1) {
 		return NULL;
 	}
+
+	MTX_LOCK(&replica->r_mtx);
 	replica->sender_epfd = epoll_create1(0);
 	event.data.fd = replica->iofd;
 	event.events = EPOLLOUT;
 	rc = epoll_ctl(replica->sender_epfd, EPOLL_CTL_ADD, replica->iofd, &event);
 	if (rc == -1) {
 		REPLICA_LOG("epoll_ctl_add failed errno:%d\n", errno);
+		MTX_UNLOCK(&replica->r_mtx);
 		return NULL;
 	}
+	MTX_UNLOCK(&replica->r_mtx);
+
 	rc = pthread_create(&replica_sender_thread, NULL, &replica_sender,
 			(void *)replica);
 	if (rc != 0) {
 		ISTGT_ERRLOG("pthread_create(replicator_thread) failed\n");
 		return NULL;
 	}
+
 	return replica;
 }
 
+static int
+send_replica_status_query(replica_t *replica, spec_t *spec)
+{
+	zvol_io_hdr_t *rmgmtio = NULL;
+	size_t data_len;
+	uint8_t *data;
+	zvol_op_code_t mgmt_opcode = ZVOL_OPCODE_REPLICA_STATUS;
+	mgmt_cmd_t *mgmt_cmd;
+
+	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+	data_len = strlen(spec->lu->volname) + 1;
+	build_replica_mgmt_hdr();
+
+	data = (uint8_t *)malloc(data_len);
+	snprintf((char *)data, BUFSIZE, "%s", spec->lu->volname);
+
+	mgmt_cmd->io_hdr = rmgmtio;
+	mgmt_cmd->io_bytes = 0;
+	mgmt_cmd->data = data;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	handle_write_data_event(replica);
+	return 0;
+}
+
+
+/*
+ * ask_replica_status will send replica_status query to all degraded replica
+ */
+static void
+ask_replica_status_all(spec_t *spec)
+{
+	int ret;
+	replica_t *replica;
+
+	MTX_LOCK(&spec->rq_mtx);
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica->state == REPLICA_HELATHY) {
+			continue;
+		}
+
+		ret = send_replica_status_query(replica, spec);
+		if (ret == -1) {
+			REPLICA_ERRLOG("send mgmtIO for status failed on "
+			    "replica(%d, %s:%d) ..\n", replica->id,
+			    replica->ip, replica->port);
+			continue;
+		}
+	}
+	MTX_UNLOCK(&spec->rq_mtx);
+}
+
+static int
+update_replica_status(spec_t *spec, replica_t *replica, zrepl_status_ack_t *repl_status)
+{
+	replica_state_t last_status;
+
+	MTX_LOCK(&replica->r_mtx);
+	last_status = replica->state;
+	replica->state = repl_status->state;
+	MTX_UNLOCK(&replica->r_mtx);
+
+	if(last_status != repl_status->state) {
+		MTX_LOCK(&spec->rq_mtx);
+		if ((replica->state == REPLICA_DEGRADED ||
+			replica->state == REPLICA_ERRORED) && last_status == REPLICA_HELATHY) {
```

Comment:
don't use replica->state here, use repl_status->state.
we don't need to use ERRORED. ERRORED state should actually get disconnected.